core/src/worker/activities.rs

```diff
@@ -509,6 +527,53 @@ where
                                     tt,
                                     ActivityCancelReason::WorkerShutdown,
                                 ));
+                            } else {
+                                // Fire off task to keep track of local timeouts. We do this so that
+                                // activities can still get cleaned up even if the user isn't
+                                // heartbeating.
+                                let local_timeout_buffer = self.local_timeout_buffer;
+                                let sched = task
+                                    .resp
+                                    .schedule_to_close_timeout
```

Comment:
I don't think schedule to close should ever come into play. I think only start to close because this is only about one attempt right? And does server set start to close on the task response even if a user didn't set it? I'm also scared of doing _anything_ with server provided times like scheduled_time. I think we need heartbeat and start-to-close only.

But maybe I didn't read/understand this algorithm close enough to understand what schedule time is for here.

Reply:
Because we use the scheduled time coming in the task, we know if this attempt would hit the overall schedule timeout - so we can handle that knowing it will have already elapsed on the server in the same way.

As for heartbeat, I got rid of it per Roey's point, b/c there's a good amount of extra machinery involved to refresh it per-heartbeat, and that seems not worth it since the only situation it matters in is when you've specified a HB timeout but aren't actually heartbeating, which would be weird.

Reply:
> Because we use the scheduled time coming in the task, we know if this attempt would hit the overall schedule timeout - so we can handle that knowing it will have already elapsed on the server in the same way.

I don't think we need to protect against that, only the attempt. I was under the impression start to close is always bound by the schedule to close time server side anyways and set here as a difference between the two.

> As for heartbeat, I got rid of it per Roey's point, b/c there's a good amount of extra machinery involved to refresh it per-heartbeat, and that seems not worth it since the only situation it matters in is when you've specified a HB timeout but aren't actually heartbeating, which would be weird.

It's not that weird, this code is very normal with a large start-to-close timeout:

```python
while True:
    activity.heartbeat()
    await my_http_client.do_hopefully_short_http_poll()
```

In fact, that's kinda the genesis of this whole issue, that users are not setting good short timeouts on HTTP calls they hoped were quick. If `do_hopefully_short_http_poll` hangs, the entire system gums up. I didn't read the logic in the PR, but ideally we can have a start to close timer that will fail task and a heartbeat timer that will fail the task, the latter simply getting reset every heartbeat call.

Reply:
There is a start to close timer. So that example still gets ended by that. But, sure, I'll see how bad it is to refresh the heartbeat and have a timer around that too

Reply:
:+1: If too much of a mess, can defer it. I wonder if the start-to-close timer is enough to never care about schedule-to-close.

Reply:
I think we want to keep schedule to close because a user might only have a schedule timeout but not a start one.

Reply:
Hrmm, I thought server _always_ set a start-to-close based on schedule-to-close if user does not explicitly have a start-to-close. But now I understand the logic, I wasn't reading ht close enough (zip, slice of two, min, etc threw me a bit)

Reply:
> I thought server always set a start-to-close based on schedule-to-close if user does not explicitly have a start-to-close

I mentioned this in one of the outdated comments, I would check what server does here and avoid this SDK complexity if possible.
I'm also a bit worried about clock skew when referencing system time, you will decrease the overall timeout if the local clock is ahead of the server clock.

Reply:
Hrmm, do we need to up that buffer leeway time from 5s to 10s, because it needs to represent the skew between server and client _and_ give the server time to timeout first?

Reply:
I added heartbeat timeouts back with resetting. Schedule doesn't really add any complexity compared to the others, so I don't see why not to keep it.